subsys/bluetooth/mesh/access.c

```diff
@@ -2354,80 +2319,253 @@ size_t comp_page_2_size(void)
+		if (!split_elems) {
+			return -EINVAL;
+		}
+		return page == 0 ? bt_mesh_metadata_get_page_0(buf, offset) : -ENOENT;
+#endif
+	default:
+		return -ENOENT;
+	}
 }
 
-int bt_mesh_comp_store(void)
+#ifdef CONFIG_BT_MESH_HIGH_DATA_PAGES
+static bool new_page_data_is_equal(enum page_type type, uint8_t page, const void *new_data,
+				   uint16_t new_len)
 {
 	NET_BUF_SIMPLE_DEFINE(buf, CONFIG_BT_MESH_COMP_PST_BUF_SIZE);
-	int err;
 
-	for (int i = 0; i < ARRAY_SIZE(comp_data_pages); i++) {
-		size_t page_size = bt_mesh_comp_page_size(i);
+	uint8_t old_page = page % 128;
+	size_t old_page_size = current_page_size(type, old_page);
 
-		if (page_size > CONFIG_BT_MESH_COMP_PST_BUF_SIZE) {
-			LOG_WRN("CDP%d is larger than the CDP persistence buffer. "
-				"Please increase the CDP persistence buffer size "
-				"to the required size (%d bytes)",
-				i, page_size);
-		}
+	if (old_page_size != new_len) {
+		return false;
+	}
 
-		net_buf_simple_reset(&buf);
+	if (old_page_size > CONFIG_BT_MESH_COMP_PST_BUF_SIZE) {
+		LOG_WRN("CDP%d is larger than the CDP persistence buffer. "
+			"Please increase the CDP persistence buffer size "
+			"to the required size (%d bytes)",
+			old_page, old_page_size);
+	}
 
-		err = bt_mesh_comp_data_get_page(&buf, comp_data_pages[i].page, 0);
-		if (err) {
-			LOG_ERR("Failed to read CDP%d: %d", comp_data_pages[i].page, err);
-			return err;
-		}
+	net_buf_simple_reset(&buf);
 
-		err = settings_save_one(comp_data_pages[i].path, buf.data, buf.len);
-		if (err) {
-			LOG_ERR("Failed to store CDP%d: %d", comp_data_pages[i].page, err);
-			return err;
-		}
+	int err = current_page_contents(&buf, type, old_page, 0, true);
 
-		LOG_DBG("Stored CDP%d", comp_data_pages[i].page);
+	if (err) {
+		LOG_ERR("Failed to read CDP%d: %d", old_page, err);
+		return false;
 	}
 
-	return 0;
+	return (memcmp(buf.data, new_data, new_len) == 0);
+}
+
+static const char *stored_page_path(enum page_type type, uint8_t page)
+{
+	for (int i = 0; i < ARRAY_SIZE(stored_pages); i++) {
+		if (stored_pages[i].type == type && stored_pages[i].page == page) {
+			return stored_pages[i].path;
+		}
+	}
+
+	return NULL;
 }
 
-int bt_mesh_comp_change_prepare(void)
+static int stored_page_write(enum page_type type, uint8_t page, const void *data, uint16_t len)
 {
+	int err;
+	/* Sentinel value used to indicate that the page is empty. */
+	uint8_t page_empty = 0;
+
 	if (!IS_ENABLED(CONFIG_BT_SETTINGS)) {
 		return -ENOTSUP;
 	}
 
-	return bt_mesh_comp_store();
+	const char *path = stored_page_path(type, page);
+
+	if (path == NULL) {
+		return -ENOENT;
+	}
+
+	/* Check that data is actually new. */
+	if (new_page_data_is_equal(type, page, data, len)) {
+		/* If page 128+n data equals page n, there is no need to store it.*/
+		data = NULL;
+	}
+
+	if (len == 0) {
+		err = settings_save_one(path, &page_empty, 1);
+	} else {
+		err = settings_save_one(path, data, data ? len : 0);
+	}
+
+	if (err) {
+		LOG_ERR("Failed to store data page: %d", err);
```

Comment:
```suggestion
		LOG_ERR("Failed to store data page %d: %d", page, err);
```

Reply:
If `page == 128` we don't know if it is metadata or comp data. I can of course expand it with `type` as well, if you think it is useful enough to add a more complex printout.

Reply:
True, could do something like. Won't insist on it, but would be nice!
```suggestion
		LOG_ERR("Failed to store %sdata page %d: %d", type == PAGE_TYPE_COMP ? "comp " : "meta", page, err);
```